### PR TITLE
#9492: replace falcon specific matmul calls

### PR DESCRIPTION
**Pull request by bbradelTT**
### Ticket
Link to Github Issue https://github.com/tenstorrent/tt-metal/issues/9492

### Problem description
- we are porting ops to ttnn

### What's changed
- there are some falcon specific matmuls that are not part of ttnn that need to be removed
- this change replaces the calls to these matmuls with generic ones in falcon models while retaining the behaviour of the specific matmuls (the specific matmuls themselves will be removed in a future PR)

### Checklist
- [x] Post commit CI passes
- [x] Model regression CI testing passes (if applicable)
- [x] New/Existing tests provide coverage for changes

https://github.com/tenstorrent/tt-metal/actions/runs/9763985533 All post-commits passes
https://github.com/tenstorrent/tt-metal/actions/runs/9763983253/job/26951645189 Nightly passes except for unstable
https://github.com/tenstorrent/tt-metal/actions/runs/9763980108 Device perf had same sporadic failure as main
https://github.com/tenstorrent/tt-metal/actions/runs/9763975702 Model perf passes
https://github.com/tenstorrent/tt-metal/actions/runs/9763956127 T3k model perf passes
https://github.com/tenstorrent/tt-metal/actions/runs/9763965001 T3k frequent passes

### CODE REVIEW

**Review by TT-BrianLiu — APPROVED** [review]
Looks good! FYI, you might need to change some expected PCC in the tests

File: models/demos/falcon7b/tests/unit_tests/test_falcon_matmuls_and_bmms_with_mixed_precision.py [hunk 1]
```diff
@@ -59,7 +62,7 @@ def run_falcon_matmul_test(
             out_mem_config = ttnn.experimental.tensor.MemoryConfig(
                 ttnn.experimental.tensor.TensorMemoryLayout.INTERLEAVED, ttnn.experimental.tensor.BufferType.DRAM
             )
-    elif falcon_op == ttnn.experimental.tensor.falcon_dense_h_to_4h_matmul:
+    elif falcon_op == "falcon_dense_h_to_4h_matmul":
```
Comment (ayerofieiev-tt) [inline models/demos/falcon7b/tests/unit_tests/test_falcon_matmuls_and_bmms_with_mixed_precision.py:65]: its a little sad to see this being changed to a string

Comment (ayerofieiev-tt) [reply models/demos/falcon7b/tests/unit_tests/test_falcon_matmuls_and_bmms_with_mixed_precision.py:65]: I feel it is totally possible to use objects

- falcon_dense_4h_to_h_matmul
- falcon_dense_h_to_4h_matmul
- falcon_lm_head_matmul

**Review by ayerofieiev-tt — APPROVED** [review]
lgtm, but see if you can get rid of strings and use an function object directly

**Review by pavlepopovic — APPROVED** [review]

**Comment by pavlepopovic** [comment]
Could you please run model_perf_regressions and device_perf_regression pipelines, we have falcon tests there?

**Comment by bbradelTT** [comment]
> Could you please run model_perf_regressions and device_perf_regression pipelines, we have falcon tests there?

Working on it. As @TT-BrianLiu said, I'll need to update the PCCs. I'm planning to increase the range to 0.02 instead of 0.01 since I'm getting values that are really close to a .01 range and am worried that randomness may just bring the value down by a fraction of a percent.

E.g.
```
Output PCC 0.900710520611985 is higher than 0.89. Please update the expected PCC
```

@pavlepopovic Is that ok?

**Comment by bbradelTT** [comment]
> > Could you please run model_perf_regressions and device_perf_regression pipelines, we have falcon tests there?
> 
> Working on it. As @TT-BrianLiu said, I'll need to update the PCCs. I'm planning to increase the range to 0.02 instead of 0.01 since I'm getting values that are really close to a .01 range and am worried that randomness may just bring the value down by a fraction of a percent.
> 
> E.g.
> 
> ```
> Output PCC 0.900710520611985 is higher than 0.89. Please update the expected PCC
> ```
> 
> @pavlepopovic Is that ok?

I talked to @pavlepopovic 
The seeds are fixed, therefore there's no randomness. I will continue to use 0.01 and update PCCs accordingly.

**Comment by bbradelTT** [comment]
> lgtm, but see if you can get rid of strings and use an function object directly

Used Enum

File: models/demos/falcon7b/tests/unit_tests/test_falcon_matmuls_and_bmms_with_mixed_precision.py (hunk 1 above)
Comment (bbradelTT) [reply models/demos/falcon7b/tests/unit_tests/test_falcon_matmuls_and_bmms_with_mixed_precision.py]: I used Enum

**Comment by bbradelTT** [comment]
> Could you please run model_perf_regressions and device_perf_regression pipelines, we have falcon tests there?

Ran both. The model perf tests have falcon tests there.

File: models/demos/falcon7b/tests/run_falcon_end_to_end.py [hunk 2]
```diff
@@ -91,12 +91,12 @@ class DeviceSetup(Enum):
         "BFLOAT16-L1": {128: (0.65, 0.77, 0.79), 1024: (0.59, 0.85, 0.85), 2047: (0.55, 0.95, 0.94)},
     },
     DeviceSetup.WORMHOLE_B0: {
-        "BFLOAT16-DRAM": {128: (0.89, 0.92, 0.92), 1024: (0.91, 0.94, 0.96), 2047: (0.96, 0.99, 0.98)},
-        "BFLOAT16-L1": {128: (0.89, 0.92, 0.92), 1024: (0.91, 0.94, 0.96), 2047: (0.96, 0.99, 0.98)},
-        "BFLOAT16-L1_SHARDED": {128: (0.89, 0.92, 0.92), 1024: (0.94, 0.95, 0.97), 2047: (0.92, 0.93, 0.95)},
+        "BFLOAT16-DRAM": {128: (0.89, 0.92, 0.91), 1024: (0.92, 0.94, 0.95), 2047: (0.95, 0.96, 0.97)},
```
Comment (uaydonat) [inline models/demos/falcon7b/tests/run_falcon_end_to_end.py:94]: Sorry, I am late to the party.

How is the new matmul "mathematically" different than the old one such that we are getting different results?

Comment (bbradelTT) [reply models/demos/falcon7b/tests/run_falcon_end_to_end.py:94]: I think the main difference is that the matmul program config parameters (per core M/N, out subblock h/w) are decided in a slightly different way, leading to a different order of operations, leading to different error accumulation sequences.

**Pull request merged**